src/mozilla_sec_eia/utils.py

```diff
@@ -78,12 +97,126 @@ def getconn() -> pg8000.dbapi.Connection:
 
     def get_metadata(self) -> pd:
         """Return dataframe of filing metadata."""
-        return pd.read_sql("SELECT * FROM sec10k_metadata", self._engine)
+        if self._metadata_df is None:
+            self._metadata_df = pd.read_sql(
+                "SELECT * FROM sec10k_metadata", self._engine
+            )
+        return self._metadata_df
 
-    def get_file(self, year_quarter: str, path: str) -> storage.Blob:
+    def get_blob(self, year_quarter: str, path: str) -> storage.Blob:
         """Return Blob pointing to file in GCS bucket."""
         return self._bucket.blob(f"sec10k/sec10k-{year_quarter}/{path}")
 
+    def _get_local_path(self, cache_directory: Path, filing: pd.Series) -> Path:
+        """Return path to a filing in local cache based on metadata."""
+        return cache_directory / Path(
+            f"{filing['CIK']}-{filing['year_quarter']}-"
+            f"{filing['Filename'].replace('edgar/data/', '').replace('/', '-')}"
+        )
+
+    def _v_stack_images(self, *images):
+        """Generate composite of all supplied images."""
+        # Get the widest width.
+        width = max(image.width for image in images)
+        # Add up all the heights.
+        height = sum(image.height for image in images)
+        composite = Image.new("RGB", (width, height))
+        # Paste each image below the one before it.
+        y = 0
+        for image in images:
+            composite.paste(image, (0, y))
+            y += image.height
+        return composite
+
+    def get_filing(
+        self,
+        filing: pd.Series,
+        cache_directory: Path = Path("./sec10k_filings"),
+    ) -> Path:
+        """Cache a single filing and return path."""
+        # Create cache directory
+        cache_directory.mkdir(parents=True, exist_ok=True)
+
+        blob = self.get_blob(filing["year_quarter"], filing["Filename"])
+        local_path = self._get_local_path(cache_directory, filing)
+
+        if exists := local_path.exists():
+            blob.update()
+            local_hash = _compute_md5(local_path)
+            remote_hash = blob.md5_hash
+            refresh = remote_hash != local_hash
+
+        if (not exists) or refresh:
+            blob.download_to_filename(local_path)
```

Comment:
Can we save the filings as .html instead of .txt? Are there any drawbacks to saving as HTMLs / do we need them as .txt filings for anything?

Reply:
I was just saving them as `.txt` because that's how the SEC distributes them, but it does seem convenient to have them as `.html` so we can open them up in a browser easier, so I've changed this